src/coreComponents/physicsSolvers/fluidFlow/StencilOutput.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file StencilOutput.cpp
+ */
+
+#include "StencilOutput.hpp"
+
+#include "common/Units.hpp"
+#include "finiteVolume/FluxApproximationBase.hpp"
+#include "constitutive/permeability/PermeabilityBase.hpp"
+#include "constitutive/permeability/PermeabilityFields.hpp"
+#include "physicsSolvers/fluidFlow/StencilAccessors.hpp"
+
+namespace geos
+{
+
+using namespace constitutive;
+using namespace dataRepository;
+
+StencilOutput::StencilOutput( const string & name,
+                              Group * const parent ):
+  Base( name, parent )
+{
+  getWrapperBase( viewKeyStruct::writeCSVFlagString() ).
+    setDescription( GEOS_FMT( "Enable the CSV output for each targeted solver discretisation scheme "
+                              "under the folder defined by {}.",
+                              Base::viewKeyStruct::writeCSVFlagString() ) );
+
+  getWrapperBase( Group::viewKeyStruct::logLevelString() ).
+    setDescription( GEOS_FMT( "When higher than 1: Display output events details.",
+                              Base::viewKeyStruct::writeCSVFlagString() ) );
+}
+
+
+bool StencilOutput::execute( real64 const time_n,
+                             real64 const dt,
+                             integer const GEOS_UNUSED_PARAM( cycleNumber ),
+                             integer const GEOS_UNUSED_PARAM( eventCounter ),
+                             real64 const GEOS_UNUSED_PARAM( eventProgress ),
+                             DomainPartition & domain )
+{
+  if( m_writeCSV > 0 && MpiWrapper::commRank() == 0 )
+  {
+    m_solver->forDiscretizationOnMeshTargets( domain.getMeshBodies(), [&] ( string const &,
+                                                                            MeshLevel & mesh,
+                                                                            arrayView1d< string const > const & )
+    {
+      NumericalMethodsManager const & numericalMethodManager = domain.getNumericalMethodManager();
+      FiniteVolumeManager const & fvManager = numericalMethodManager.getFiniteVolumeManager();
+      FluxApproximationBase const & fluxApprox = fvManager.getFluxApproximation( m_solver->getDiscretizationName() );
+
+      fluxApprox.forStencils< CellElementStencilTPFA >( mesh, [&]( auto const & stencil )
```

Comment:
could you please check if all that will work for https://github.com/GEOS-DEV/GEOS/blob/develop/inputFiles/compositionalMultiphaseFlow/co2_hybrid_1d.xml ?
not sure hybrid discretization has the same things as TPFA

Reply:
I suppose the key here is that `fvManager.getFluxApproximation( m_solver->getDiscretizationName() );` might return null if the solver doesn't have a flux approximation. So I would add a validator (`postProcessInput()`) to the `solverName` field of this to ensure that the solver provided has the appropriate flux approximation.